proposals/0000-quick-look-impredicativity.md

```diff
@@ -0,0 +1,181 @@
+
+```text
+Prelude> :set -XRankNTypes -XImpredicativeTypes
+Prelude> choose :: (a -> Bool) -> (a -> Bool) -> Bool ; choose = undefined
+Prelude> f :: (Bool -> Bool) -> Bool ; f = undefined
+Prelude> g :: (forall a. a -> a) -> Bool ; g = undefined
+Prelude> :t choose f g
+
+<interactive>:1:10: error:
+    • Couldn't match type ‘Bool -> Bool’ with ‘forall a. a -> a’
+      Expected type: (Bool -> Bool) -> Bool
+        Actual type: (forall a. a -> a) -> Bool
+    • In the second argument of ‘choose’, namely ‘g’
+      In the expression: choose f g
+Prelude> :t choose g f
+choose g f :: Bool
+```
+
+There is a simple workaround that works most of the time: eta-expansion (you might need to eta-expand more than once, though). Taking the `choose`, `f`, and `g` functions defined above,in a non-contravariant world we need to eta-expand `f` to make it work: `choose (\x -> f x) g`; now the `x` is assigned type `forall a. a -> a`, which is then instantiated to `Int -> Int` before passing it to `f`.
+
+### The `ContravariantFunctions` extension
+
+After a preliminary assessment of which packages still compile under this new policy, it looks like most of them do not need it. Except one big and important one: `Cabal` (and anything that uses its API). This library often uses functions of the form:
+
+```haskell
+f :: ( (... -> CabalIO a) -> ...) -> ...
+```
+
+where `CabalIO a` is defined as `HasCallStack => IO a`. That means that whereas before we could pass a function `g` as argument, now we have to do something akin to `\... t ... -> g (\... x ... -> t ... x ...)`, which is quite annoying.
+
+We thus propose to add another extension to GHC, called `ContravariantFunctions`, which should be disabled by default. When the extension is enabled, subsumption checking is done using full contravariance of function types (but during quick look contravariance does not enter the game). This means that `RankNTypes` + `ContravariantFunctions` behaves the same as the old `RankNTypes. This has the benefit of asking people to eta-expand their functions, while having a escape way for those libraries which would require too many changes.
+
+### The typing rule for `($)`
+
+Currently, GHC contains a hard-coded typing rule for `($)` (you can find it in [this file](https://gitlab.haskell.org/ghc/ghc/blob/master/compiler/typecheck/TcExpr.hs), looking for `dollarIdKey`) which ensures that `f $ e` works even when `($)` would need to be impredicatively instantiated. The million dollar question is: can we drop this special case from the compiler?
```

Comment:
Here is a permalink to the typing rule. It's probably better than asking the reader to search: https://gitlab.haskell.org/ghc/ghc/blob/795986aaf33e2ffc233836b86a92a77366c91db2/compiler/typecheck/TcExpr.hs#L368-397